### ISSUE TITLE
Regent: bounds checks on specific task only

### ISSUE THREAD

**Issue by manopapad** [StanfordLegion/legion#578]
This came in handy while chasing down a bug in Soleil that was deterministically causing a CUDA memory error, but only after 700 iterations. Adding bounds checks to all tasks would have made the run take forever, but it was only a small overhead to add them just to the offending task. @magnatelee hacked the compiler to achieve this, but it might be nice to expose this functionality to the user.

**Comment by magnatelee** [comment]
Commit f4f119c9 adds the flag `-fbounds-checks-targets` to the compiler. The flag takes a comma-separate list of regex-like patterns that specify the target of bounds checks (the default pattern is `.*`). Lua's built-in support for regex is not as expressive as regex in other languages (for example, `(abc|def)` is not allowed), so I had to compensate that by allowing a list of patterns. @manopapad Please close this issue once you check out the new flag.

**Comment by streichler** [comment]
@magnatelee how hard would it be to use the posix regex suport (regcomp, regexec) instead of Lua's builtin-but-inferior regular expressions?

**Comment by manopapad** [comment]
AFAIK invoking external libraries at Regent compilation time requires either writing a wrapper for them using the [Lua C API](https://www.lua.org/manual/5.2/manual.html#4), which is a lot of boilerplate, or invoking them through luajit, which automatically handles the calling convention between Lua and C, but requires that luajit be working properly on the target platform (not the case e.g. for power9). I think we have been trying to avoid relying on luajit as much as possible at compile time (there is even a built configuration for Regent that avoids using luajit altogether, if all you're doing is compiling to object files).

Long story short, using posix regex at compile time is possible but either requires too much code for anyone to want to do it, or forces us to go down a path we don't want to be going down (my 2 cents, @elliottslaughter  or @magnatelee might have more insight into this).

**Comment by streichler** [comment]
Yes, my suggestion assumed we had terra goodness available to us at compile time.  If we do not, then living with lua's regex capabilities seems the more pragmatic approach.

**Comment by magnatelee** [comment]
Using a more powerful regex library in C, either with Terra or some third-party libraries like Lrexlib, is not hard but, as both of you pointed out, not portable either. I expect that most users would just want to specify several tasks that they want to run bounds checks on, so I don't think using Lua's inferior pattern support isn't actually losing that much. (Or in other words, adding more powerful regex support seems like just a hassle without any real gain.)

**Comment by manopapad** [comment]
I've confirmed that this works, by playing around with the following example:

```
import 'regent'

task init(r : region(ispace(int1d),double))
where writes(r) do
  for i in r do
    r[i] = 1.0
  end
end

task foo(r : region(ispace(int1d),double))
where reads writes(r) do
  for i in r do
    r[i] += r[i+1]
  end
end

task main()
  var r = region(ispace(int1d,10), double)
  init(r)
  var p = partition(equal, r, ispace(int1d,2))
  foo(p[0])
  for i in r do
    regentlib.c.printf('%lf\n', r[i])
  end
end
regentlib.start(main)
```

The call to `foo(p[0])` performs an out-of-bounds access, but this won't manifest as a segfault in a single-node execution, as the default mapper will reuse the full-region instance created for `init` when mapping `foo`.
